### ISSUE TITLE
Is it possible to precompile for mips and mipsel?

### ISSUE THREAD

**Issue by christaikobo** [zephyrchien/kaminari#5]
First of all, thanks for this project, great work!

It seems that the binary files size is quite small.

I am thinking about running these on a router if possible.

**Comment by zephyrchien** [comment]
Unfortunately not, If you need tls support. Because `rustls` depends on `ring`, which does not compile on mips.

I'm going to add build options to disable tls support, and add `openssl` as alternative tls backend.

**Comment by christaikobo** [comment]
Thanks for the swift response!

Also, I am having trouble understanding mask mode, can you please explain quickly the pros and cons of each mode? And, will mask mode be compatible with v2ray-plugin client? Thank you!

**Comment by zephyrchien** [comment]
A websocket client will **mask(xor)** its data with a 32bit key, and send the key together with data. The receiver(server/middleware) will obtain the key, and **unmask(xor)** the data. 

`mode` here only affects how the masking key is generated, **so theorily kaminaric could communicate with any other  websocket server**, whatever `mode` you use. 

According to [rfc6455](https://www.rfc-editor.org/rfc/rfc6455.html#section-5.3),
>When preparing a masked frame, the client MUST pick a fresh masking
   key from the set of allowed 32-bit values.  The masking key needs to
   be unpredictable; thus, the masking key MUST be derived from a strong
   source of entropy, and the masking key for a given frame MUST NOT
   make it simple for a server/proxy to predict the masking key for a
   subsequent frame.

Please consider `mode=standard` as standard bahavior.

For `mode=fix`, the key will not be regenerated unless there comes another connection, so that a few CPU cycles could be saved(this is meaningless, tbh), **please do not use this mode**.

For `mode=skip`, it is a dirty(?) hack invented by me. With a key = 0x00..0(32bit), the mask(xor) operation does nothing, and could be skipped. **Our server(kaminaris) will also skip unmasking when receiving an empty key. Other servers will do xor
(data, 0x00..0) as normal**, which leads to the same result but costs more CPU cycles.

In conclusion, if you prefer a more standard implementation, please use `mode=standard`; if you prefer performance, please use `mode=skip`. I personally prefer the later, this is why I set it as the default mode.

**Comment by zephyrchien** [comment]
On the other hand, when I was developing `lightws`, ~~the underlying library of `kaminari`, which implements the websocket protocol(rfc6455)~~ at the beginning there was only `mode=skip`, while `mode=fix` and `mode=standard` are added later.

The main reason is that the `Write` trait takes an immutable buffer reference, where there is no way to mask it unless I bring in some unsafe code which will break rust's borrowing rules and should be carefully used. The clever solution is simply skip this step:). For more implementation details, read [lightws's document](https://docs.rs/lightws/latest/lightws/stream/index.html#masking-payload), if you are interested.

**Comment by christaikobo** [comment]
Wow, really appreciate the detailed explanation! 

So I guess there is not much of a difference in terms of circumventing censorship between mask modes. It is really just down to implementing rfc6455.

**Comment by christaikobo** [comment]
I know it has been forever, is there any plan for this?

**Comment by zephyrchien** [comment]
@christaikobo Thank your for the continued interest in this project and sorry for the fact that I'm not able to spare a lof of time on it. It has been 3 years and a lot of things changed. For example, there comes [aws/s2n-tls](https://github.com/aws/s2n-tls) which works on mips and seems easier to integrate with.

**Comment by christaikobo** [comment]
Thanks for the update, are you interested in implementing it? If not I think it's safe to close this issue.

**Comment by zephyrchien** [comment]
See https://docs.rs/crate/rustls/latest, I think we just need to recompile rustls with a different provider, like aws-ls or openssl.

**Comment by zephyrchien** [comment]
The latest kaminari uses `aws-lc2` as `rustls` backend, it should be able to work on more platforms.

**Comment by christaikobo** [comment]
Thanks, there does not seem to be prebuilt binaries?

**Comment by zephyrchien** [comment]
@christaikobo  I added mips builds to [workflow](https://github.com/zephyrchien/kaminari/actions/runs/16860808320/job/47760687271) and most of them worked fine, except for `mips64el-unknown-linux-muslabi64` which I'm trying to fix. If you're interested you could have a try with other mips builds, hope them work for you.

**Comment by christaikobo** [comment]
Just tried the mipsel build, it's working!

Thanks!
